### ISSUE TITLE
In install.php, if database exists, the installer stops

### ISSUE THREAD

**Issue by MagicFab** [bemosior/PHPasswordPusher#34]
The installer may not assume the database doesn't already exist. It currently stops if it doesn't.

**Comment by bemosior** [comment]
We could probably do more to make the install process idempotent. Any suggestions?

**Comment by MagicFab** [comment]
I have no clue what you mean in this context (and it's 1AM) :) Could you rephrase?

**Comment by bemosior** [comment]
"Idempotent" in this case would mean being able to run the install process multiple times with the same parameters and not have an alternate outcome. Your example for this issue is: Even if the database exists, still perform the remaining installation. We would want to take that approach for each step in the installation process.

I think we'd end up leveraging the `IF NOT EXISTS` part described in the [documentation](https://dev.mysql.com/doc/refman/5.0/en/create-database.html) for the `CREATE DATABASE` statement. Of course, we would want that same kind of logic (adjusted appropriately) for each step in the install process.

**Comment by MagicFab** [comment]
I don't expect the tables to already exist but I think it's fair to continue after a proper database/permissions check. In the end it's just validation, if everything needed already exists, then the install becomes a validation checklist. This issue report would be about the database being properly setup, otherwise creating it and advising in a proper message when/if something is missing for this to happen.

**Comment by bemosior** [comment]
Installer as Validator is easy to get behind, and idempotence is a prerequisite. It's probably good timing, too. I'd like to be able to add support for Postgres, so making each step parameterizable is best.

**Comment by MagicFab** [comment]
Perhaps open a separate ticket as feature request or enhacement for PG or similar.

**Comment by vaindil** [comment]
a734529 only partially addresses this, shouldn't have closed this.

**Comment by vaindil** [comment]
Due to the nature of the of the project, I think the proper way to proceed would be with the change I started, but to check the DB's/table's configuration rather than just dropping them if they already exist. (Checking and only making the necessary changes could break things for previous users. I'll write something in the documentation or a wiki page about it; leaving this open until then.)

**Comment by vaindil** [comment]
Actually, after reviewing the available options in PHP (primarily [`getColumnMeta`](https://secure.php.net/manual/en/pdostatement.getcolumnmeta.php)), it seems the best way would just be to check the database/table info as in a734529. There's no guarantee that a specific version of MySQL or a particular driver would return the same results to allow for proper verification each time. I'll leave the issue open for now in case others have suggestions.

**Comment by bemosior** [comment]
@vaindil I think your changes maintain a good balance between value and effort investment. For the Laravel rewrite, we can take advantage of a more standard database setup mechanism (migrations, etc.) that doesn't have these issues (and supports MySQL, PostgreSQL, SQLite, and SQL Server). Thanks for your work on this. :)
